- открыть https://rdrr.io/snippets/

- вставить код (поменять сид на свой)

In [ ]:
set.seed(70527)
X1 <- round(rnorm(100), 10)
X2 <- round(rnorm(100, mean = 5), 2)
X3 <- round(runif(100, min = 5, max = 20), 2)
b0 <- round(runif(1, 10, 20), 0)
b1 <- round(runif(1, 30, 50), 0)
b2 <- round(runif(1, 30, 50), 0)
b3 <- round(runif(1, 30, 50), 0)
Y <- b0 + b1 * X1 + b2 * X2 + b3 * X3 + rnorm(100, 0, 100)
Y_bin <- b2 + X1 - X2 - X3 + rnorm(100, 0, 5)
Y_bin <- as.numeric(Y_bin > mean(Y_bin))
myData <- data.frame(Y, Y_bin, X1, X2, X3)
myData

- скопировать получившийся дф (100 строк должно быть)

- вставить в код ниже и сделать пандас датафрейм

In [8]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

In [ ]:
data = [
    [1, 642.9645, 0, -0.1902519275, 5.30, 13.19],
    [2, 669.0785, 0, 1.1587101475, 5.27, 12.83],
    [3, 757.3738, 1, -0.7083306134, 4.99, 16.23],
    [4, 207.0491, 1, 1.0122087225, 3.87, 5.65],
    [5, 609.1017, 1, 0.7552243444, 5.57, 12.68],
    [6, 434.7039, 1, 0.3882895931, 5.37, 12.82],
    [7, 555.2668, 0, -0.8868628470, 5.04, 19.13],
    [8, 320.7739, 1, -0.5428500754, 6.11, 5.70],
    [9, 641.8875, 0, -0.8322325836, 5.52, 11.84],
    [10, 305.6080, 1, 0.1874712633, 4.80, 8.57],
    [11, 767.3438, 0, -0.7466860216, 6.84, 17.18],
    [12, 438.8108, 0, -1.7668190822, 5.58, 9.95],
    [13, 649.7338, 0, -0.8700010689, 4.39, 17.44],
    [14, 721.4392, 0, -1.3356172102, 6.67, 14.26],
    [15, 472.4274, 1, -0.2052322567, 5.56, 11.69],
    [16, 350.9432, 1, 0.2000589690, 4.75, 8.95],
    [17, 481.3434, 0, 0.8945026100, 5.02, 6.98],
    [18, 620.3198, 0, 0.2293359648, 5.92, 15.60],
    [19, 543.0805, 1, -0.9086235250, 5.27, 10.81],
    [20, 840.9555, 0, 1.4189463370, 6.45, 17.01],
    [21, 682.6461, 0, -0.7782509431, 5.80, 16.64],
    [22, 248.4116, 1, -0.9850683205, 4.28, 8.77],
    [23, 608.7741, 0, -0.2840474416, 5.60, 15.85],
    [24, 290.6755, 1, 0.1297600130, 5.10, 7.18],
    [25, 565.2029, 0, -0.3398579104, 5.71, 13.06],
    [26, 414.3131, 1, -0.0006763375, 5.95, 13.99],
    [27, 278.5390, 1, -0.3407329187, 3.76, 5.46],
    [28, 558.5398, 0, -0.7553372612, 5.96, 17.35],
    [29, 898.4615, 1, 0.8551903073, 4.15, 16.04],
    [30, 561.8427, 1, -1.4263149720, 6.39, 9.65],
    [31, 844.9739, 0, -0.7950917307, 4.16, 15.83],
    [32, 738.1746, 0, -0.8262846679, 4.16, 15.93],
    [33, 337.1722, 1, 1.4224859389, 6.15, 5.19],
    [34, 758.0764, 0, -1.7727784005, 5.35, 14.56],
    [35, 349.1956, 1, 0.7403880176, 3.88, 11.22],
    [36, 714.8325, 0, 1.9428861853, 3.48, 15.72],
    [37, 458.3363, 0, -1.3273568528, 5.98, 13.64],
    [38, 500.6815, 0, -0.3039161674, 2.85, 12.87],
    [39, 613.0279, 0, -0.9653243073, 7.14, 13.66],
    [40, 678.9514, 0, 1.2686319705, 5.51, 14.08],
    [41, 801.4217, 0, 0.2451104394, 8.05, 18.63],
    [42, 658.7345, 1, 0.2476842213, 5.15, 10.10],
    [43, 333.2191, 1, -0.2145810160, 3.72, 7.21],
    [44, 508.6670, 1, 0.7071614258, 6.22, 13.59],
    [45, 400.0633, 1, -0.4713798181, 5.34, 9.85],
    [46, 625.0702, 0, 0.0363763475, 5.70, 14.82],
    [47, 432.7317, 1, -0.9040922031, 4.30, 9.57],
    [48, 939.3007, 0, 1.9729284746, 5.49, 18.19],
    [49, 710.7056, 1, -1.0536991667, 4.41, 17.77],
    [50, 802.1650, 0, -1.5188071464, 6.86, 19.44],
    [51, 432.4508, 0, 1.4325569483, 6.21, 7.55],
    [52, 553.8942, 0, -0.1638500018, 4.80, 17.77],
    [53, 489.9851, 1, -0.0697794247, 3.99, 14.70],
    [54, 648.5112, 0, -0.7805991828, 6.42, 15.45],
    [55, 602.3947, 0, 0.0155355942, 4.86, 14.99],
    [56, 314.9670, 1, 0.1824372498, 5.15, 5.49],
    [57, 880.5772, 0, 1.5682367683, 4.87, 18.73],
    [58, 571.1068, 1, -0.2241986079, 5.26, 13.00],
    [59, 505.9300, 0, -2.1060470022, 6.67, 17.24],
    [60, 369.2822, 1, -0.8225706324, 4.21, 10.49],
    [61, 330.7144, 1, -0.5689675953, 3.49, 10.21],
    [62, 251.0186, 1, -0.2402111526, 4.55, 9.36],
    [63, 509.2846, 0, -1.3321501879, 5.22, 15.16],
    [64, 683.7051, 0, 0.7479081343, 5.57, 10.62],
    [65, 963.1007, 0, 0.1427611269, 4.36, 19.95],
    [66, 454.1292, 0, -1.1381329507, 4.77, 14.20],
    [67, 649.7590, 0, 0.4448685697, 3.95, 16.47],
    [68, 633.4207, 1, 0.5839670383, 3.40, 10.88],
    [69, 568.4918, 0, 0.5672352175, 3.46, 17.56],
    [70, 380.4988, 1, 0.5149030065, 4.04, 6.02],
    [71, 455.5337, 1, 0.5069535060, 6.10, 10.31],
    [72, 642.0871, 1, 0.3711830680, 6.25, 18.14],
    [73, 619.8503, 0, 0.5948570251, 4.67, 14.10],
    [74, 404.7538, 1, -0.4301078717, 5.37, 13.56],
    [75, 612.6624, 1, -0.4463268921, 4.44, 9.41],
    [76, 653.9840, 0, -0.6602043603, 4.05, 16.47],
    [77, 867.3363, 1, 3.0420567855, 5.37, 15.30],
    [78, 687.4651, 0, 0.8735182190, 4.54, 13.97],
    [79, 844.9099, 0, 0.3105628956, 5.83, 18.90],
    [80, 514.1101, 1, -0.0620273546, 5.69, 14.31],
    [81, 548.9489, 1, -1.1869061419, 3.85, 10.49],
    [82, 556.9445, 1, 0.2695415502, 5.28, 13.82],
    [83, 1056.0112, 0, 1.9436222398, 6.66, 18.78],
    [84, 583.9669, 0, -0.5057365427, 3.47, 19.51],
    [85, 670.8295, 0, 0.8606486568, 6.33, 12.87],
    [86, 325.8764, 1, -0.2747131830, 4.08, 5.38],
    [87, 752.1300, 0, -0.7755235970, 6.17, 19.77],
    [88, 810.4297, 0, 0.3710574702, 5.09, 17.17],
    [89, 506.9250, 1, -0.2510339471, 5.07, 10.49],
    [90, 742.4601, 1, 0.5174211435, 3.52, 14.93],
    [91, 728.2394, 0, 0.3135168060, 6.67, 12.67],
    [92, 661.7737, 1, -0.7455363354, 5.36, 14.22],
    [93, 616.0665, 0, -0.1605178821, 5.06, 19.52],
    [94, 435.6907, 1, 0.1642344469, 6.09, 5.71],
    [95, 410.1606, 1, 0.7237035197, 2.84, 14.17],
    [96, 689.1581, 0, -0.6345071188, 5.25, 18.20],
    [97, 325.4261, 1, 1.0405991790, 3.35, 9.27],
    [98, 402.9967, 1, 0.4892772883, 4.31, 7.42],
    [99, 563.0867, 0, 0.5302531792, 4.69, 17.67],
    [100, 399.1970, 1, -1.1883363023, 5.01, 13.36],
]

df = pd.DataFrame(data, columns=["index", "Y", "Y_bin", "X1", "X2", "X3"]).set_index("index")
df.index.name = None
df

,Y,Y_bin,X1,X2,X3
1,642.9645,0,-0.190252,5.30,13.19
2,669.0785,0,1.158710,5.27,12.83
3,757.3738,1,-0.708331,4.99,16.23
4,207.0491,1,1.012209,3.87,5.65
5,609.1017,1,0.755224,5.57,12.68
...,...,...,...,...,...
96,689.1581,0,-0.634507,5.25,18.20
97,325.4261,1,1.040599,3.35,9.27
98,402.9967,1,0.489277,4.31,7.42
99,563.0867,0,0.530253,4.69,17.67


---
# Задача 1
### Оцени регрессию Y на X1, X2, X3 и куб переменной X2. В ответ введи R^2 этой регрессии

Рассматривается модель множественной линейной регрессии:

$$
Y_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i + \beta_4 X2_i^3 + \varepsilon_i
$$

### Теория

Это модель, линейная по параметрам, поэтому её можно оценить методом наименьших квадратов (МНК).

Согласно теореме Гаусса–Маркова, если выполнены условия:

- модель правильно специфицирована  
- отсутствует линейная зависимость между регрессорами  
- $E(\varepsilon_i) = 0$  
- $Var(\varepsilon_i) = \sigma^2$  
- $Cov(\varepsilon_i, \varepsilon_j) = 0$ при $i \ne j$  

то оценки МНК являются BLUE (наилучшими линейными несмещёнными оценками) :contentReference[oaicite:0]{index=0}  

Добавление переменной $X2^3$ позволяет учесть нелинейную зависимость между $Y$ и $X2$, при этом модель остаётся линейной по коэффициентам.

### Коэффициент детерминации

$$
R^2 = 1 - \frac{RSS}{TSS}
$$

где:
- $RSS = \sum \hat{\varepsilon}_i^2$ — сумма квадратов остатков  
- $TSS = \sum (Y_i - \bar{Y})^2$ — общая сумма квадратов  

$R^2$ показывает долю вариации зависимой переменной $Y$, объяснённую моделью.

In [ ]:
df["X2_cubed"] = df["X2"] ** 3

X = df[["X1", "X2", "X3", "X2_cubed"]]
X = sm.add_constant(X)
y = df["Y"]

model = sm.OLS(y, X).fit()

model.rsquared

np.float64(0.6419192769598947)

### Интерпретация

Это означает, что примерно **64.2% вариации зависимой переменной $Y$ объясняется моделью**, включающей переменные $X1$, $X2$, $X3$ и нелинейный член $X2^3$.

Оставшиеся **35.8% вариации** обусловлены:
- неучтёнными факторами,
- случайной составляющей $\varepsilon_i$,
- возможной неправильной спецификацией модели.


---
# Задача 2
### Рассмотрим три модели: модель 1 — это регрессия $Y$ на $X1$, $X2$, $X3$ и куб $X1$; модель 2: $Y$ на $X1$, $X2$, $X3$ и куб $X2$; модель 3: $Y$ на $X1$, $X2$, $X3$ и куб $X3$. В какой из этих моделей достигается максимальное значение скорректированного $R^2$?

Нужно сравнить три модели по **скорректированному коэффициенту детерминации**:

$$
\bar{R}^2 = 1 - \frac{RSS/(n-k-1)}{TSS/(n-1)}
$$

где:
- $RSS$ — сумма квадратов остатков,
- $TSS$ — общая сумма квадратов,
- $n$ — число наблюдений,
- $k$ — число регрессоров без константы.

### Зачем нужен скорректированный $R^2$

Обычный $R^2$ не уменьшается при добавлении новых регрессоров, даже если они почти бесполезны. Поэтому для сравнения моделей с разными дополнительными переменными удобнее использовать именно $\bar{R}^2$: он учитывает не только качество подгонки, но и штрафует модель за избыточную сложность.

### Что делаем в задаче

Оцениваем три регрессии:

$$
\text{Модель 1: } Y_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i + \beta_4 X1_i^3 + \varepsilon_i
$$

$$
\text{Модель 2: } Y_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i + \beta_4 X2_i^3 + \varepsilon_i
$$

$$
\text{Модель 3: } Y_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i + \beta_4 X3_i^3 + \varepsilon_i
$$

После этого сравниваем значения $\bar{R}^2$ и выбираем модель, где оно максимально.

In [4]:
df["X1_cubed"] = df["X1"] ** 3
df["X2_cubed"] = df["X2"] ** 3
df["X3_cubed"] = df["X3"] ** 3

y = df["Y"]

X1_model = sm.add_constant(df[["X1", "X2", "X3", "X1_cubed"]])
X2_model = sm.add_constant(df[["X1", "X2", "X3", "X2_cubed"]])
X3_model = sm.add_constant(df[["X1", "X2", "X3", "X3_cubed"]])

model_1 = sm.OLS(y, X1_model).fit()
model_2 = sm.OLS(y, X2_model).fit()
model_3 = sm.OLS(y, X3_model).fit()

adj_r2 = {
    1: model_1.rsquared_adj,
    2: model_2.rsquared_adj,
    3: model_3.rsquared_adj,
}

print(adj_r2)
print(max(adj_r2, key=adj_r2.get))

{1: np.float64(0.6332035244667489), 2: np.float64(0.6268421938845219), 3: np.float64(0.6279354126079808)}
1


### Интерпретация

Максимальное значение скорректированного коэффициента детерминации достигается в **модели 1**:

$$
\bar{R}^2_1 = 0.633 > \bar{R}^2_3 = 0.628 > \bar{R}^2_2 = 0.627
$$

Это означает, что именно модель с добавлением $X1^3$ **лучше всего объясняет вариацию $Y$ с учётом штрафа за сложность модели**.

Следовательно, нелинейность по переменной $X1$ является наиболее важной среди рассмотренных вариантов, тогда как добавление кубов $X2$ или $X3$ даёт менее качественное улучшение модели.

---
# Задача 3
### Рассмотрим регрессию $Y$ на $X1$, $X2$ и $X3$. Какая получится $F$-статистика при проверке гипотезы о совместной незначимости $X1$ и $X2$ в этой модели?

Рассматривается модель множественной линейной регрессии:

$$
Y_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i + \varepsilon_i
$$

Нужно проверить гипотезу о **совместной незначимости** переменных $X1$ и $X2$:

$$
H_0: \beta_1 = \beta_2 = 0
$$

$$
H_1: \text{хотя бы один из коэффициентов } \beta_1, \beta_2 \neq 0
$$

### Теория

Для проверки линейных ограничений на коэффициенты регрессии используется $F$-тест. В лекциях отмечается, что $F$-тест применяется именно для проверки гипотез о **линейных ограничениях на коэффициенты регрессии** :contentReference[oaicite:0]{index=0}

В этой задаче нулевая гипотеза означает, что после учёта $X3$ переменные $X1$ и $X2$ **совместно не вносят вклад** в объяснение $Y$.

Практически это делается через сравнение:
- **неограниченной модели**:
$$
Y_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i + \varepsilon_i
$$

- **ограниченной модели** при $H_0$:
$$
Y_i = \beta_0 + \beta_3 X3_i + \varepsilon_i
$$

Если $F$-статистика велика, то ограничение слишком сильно ухудшает качество модели, и гипотеза $H_0$ отвергается.

In [5]:
y = df["Y"]

X_ur = sm.add_constant(df[["X1", "X2", "X3"]])
X_r = sm.add_constant(df[["X3"]])

model_ur = sm.OLS(y, X_ur).fit()
model_r = sm.OLS(y, X_r).fit()

f_stat, p_value, df_diff = model_ur.compare_f_test(model_r)
f_stat

np.float64(10.98014439542532)

### Интерпретация

Полученная статистика

$$
F = 10.98
$$

используется для проверки гипотезы

$$
H_0: \beta_1 = \beta_2 = 0
$$

против альтернативы

$$
H_1: \text{хотя бы один из коэффициентов } \beta_1, \beta_2 \neq 0
$$

Такое значение $F$ является довольно большим, поэтому совместная гипотеза о незначимости $X1$ и $X2$, скорее всего, будет отвергаться на стандартных уровнях значимости.

Это означает, что переменные $X1$ и $X2$ **совместно значимы** в регрессии $Y$ на $X1$, $X2$ и $X3$: после учёта $X3$ они всё ещё добавляют объясняющую силу модели.

---
# Задача 4
### Создай переменную $D$, равную 1, если $X3$ больше своего среднего, и 0 иначе. Оцени регрессию $Y$ на $D$, в ответ введи коэффициент при $D$ в этой регрессии

Рассматривается модель:

$$
Y_i = \beta_0 + \beta_1 D_i + \varepsilon_i
$$

где $D_i$ — фиктивная (dummy) переменная:

$$
D_i =
\begin{cases}
1, & \text{если } X3_i > \bar{X3} \\
0, & \text{иначе}
\end{cases}
$$

### Теория

Такая регрессия называется моделью с бинарной (фиктивной) переменной.

В этом случае коэффициенты имеют простую интерпретацию:

- $\beta_0 = E(Y \mid D=0)$ — среднее значение $Y$ в группе, где $X3 \le \bar{X3}$  
- $\beta_1 = E(Y \mid D=1) - E(Y \mid D=0)$ — разница средних между группами  

То есть коэффициент при $D$ показывает, **насколько в среднем отличается $Y$ для наблюдений с высоким $X3$ относительно низкого $X3$**.

In [6]:
df["D"] = (df["X3"] > df["X3"].mean()).astype(int)

y = df["Y"]
X = sm.add_constant(df[["D"]])

model = sm.OLS(y, X).fit()
model.params["D"]

np.float64(207.97956980519456)

### Интерпретация

Оценённый коэффициент при фиктивной переменной равен

$$
\hat{\beta}_1 = 207.98
$$

Это означает, что для наблюдений, у которых $X3$ **выше среднего значения**, среднее значение $Y$ в среднем **на 207.98 единиц больше**, чем для наблюдений, у которых $X3$ не превышает среднее.

Иными словами, переменная $D$ делит выборку на две группы, и коэффициент при $D$ показывает разницу средних значений $Y$ между этими группами.

---
# Задача 5
### Найди максимальную по модулю корреляцию между объясняющими переменными ($X1$, $X2$, $X3$). В ответ введи модуль этой корреляции

### Теория

Для анализа зависимости между объясняющими переменными используется **коэффициент корреляции Пирсона**:

$$
\rho_{X_j, X_k} = \frac{Cov(X_j, X_k)}{\sqrt{Var(X_j) Var(X_k)}}
$$

Он принимает значения от $-1$ до $1$:
- $|\rho| \approx 1$ — сильная линейная зависимость  
- $|\rho| \approx 0$ — слабая линейная зависимость  

В контексте эконометрики высокая корреляция между регрессорами может свидетельствовать о **мультиколлинеарности** :contentReference[oaicite:0]{index=0}  

В задаче нужно:
1. Построить корреляционную матрицу для $X1$, $X2$, $X3$  
2. Найти максимальное значение по модулю среди попарных корреляций

In [9]:
corr_matrix = df[["X1", "X2", "X3"]].corr()

corr_matrix.abs().where(~np.eye(3, dtype=bool)).max().max()

np.float64(0.183242333024623)

### Интерпретация

Максимальная по модулю корреляция между объясняющими переменными равна

$$
|\rho| = 0.183
$$

Это небольшое значение, что свидетельствует о **слабой линейной зависимости между регрессорами**.

Следовательно, в модели **нет проблемы мультиколлинеарности**, и оценки коэффициентов МНК не должны быть нестабильными из-за сильной корреляции между объясняющими переменными.

---
# Задача 6
### В регрессии $Y$ на $X1$, $X2$, $X3$ найди VIF для переменной $X2$

### Теория

Фактор инфляции дисперсии (Variance Inflation Factor, VIF) используется для диагностики мультиколлинеарности.

Для переменной $X_j$:

$$
VIF_j = \frac{1}{1 - R_j^2}
$$

где $R_j^2$ — коэффициент детерминации из регрессии переменной $X_j$ на все остальные объясняющие переменные.

В нашем случае:

$$
X2_i = \gamma_0 + \gamma_1 X1_i + \gamma_2 X3_i + u_i
$$

и

$$
VIF_{X2} = \frac{1}{1 - R^2_{X2}}
$$

### Интерпретация VIF

- $VIF = 1$ — отсутствует корреляция с другими регрессорами  
- $VIF > 5$ — умеренная мультиколлинеарность  
- $VIF > 10$ — сильная мультиколлинеарность :contentReference[oaicite:0]{index=0}  

Таким образом, VIF показывает, насколько дисперсия оценки коэффициента увеличивается из-за корреляции с другими переменными.

In [10]:
X_aux = sm.add_constant(df[["X1", "X3"]])
y_aux = df["X2"]

model_aux = sm.OLS(y_aux, X_aux).fit()
1 / (1 - model_aux.rsquared)

np.float64(1.0420082138418019)

### Интерпретация

Полученное значение

$$
VIF_{X2} = 1.042
$$

очень близко к 1, что означает практически полное отсутствие линейной зависимости между $X2$ и другими регрессорами ($X1$, $X3$).

Следовательно, **мультиколлинеарность для переменной $X2$ отсутствует**, и дисперсия оценки её коэффициента практически не увеличивается из-за корреляции с другими переменными.

---
# Задача 7
### Оцени регрессию логарифма $Y$ на логарифм $X3$. В ответ введи значение коэффициента при логарифме $X3$

Рассматривается модель:

$$
\ln Y_i = \beta_0 + \beta_1 \ln X3_i + \varepsilon_i
$$

### Теория

Это **лог-лог модель**.

Коэффициент $\beta_1$ интерпретируется как эластичность:

$$
\beta_1 = \frac{\partial \ln Y}{\partial \ln X3}
$$

Это означает, что:
- при увеличении $X3$ на 1%  
- $Y$ в среднем изменяется на $\beta_1$\%

Такая спецификация часто используется, когда предполагается **нелинейная (степенная) зависимость** между переменными.

In [11]:
df["lnY"] = np.log(df["Y"])
df["lnX3"] = np.log(df["X3"])

y = df["lnY"]
X = sm.add_constant(df[["lnX3"]])

model = sm.OLS(y, X).fit()
model.params["lnX3"]

np.float64(0.7328244971735387)

### Интерпретация

Оценённый коэффициент равен

$$
\hat{\beta}_1 = 0.733
$$

Это означает, что при увеличении $X3$ на 1% зависимая переменная $Y$ в среднем увеличивается примерно на **0.733%**.

Таким образом, зависимость между $Y$ и $X3$ является положительной и менее чем пропорциональной (неэластичной), так как коэффициент меньше 1.

---
# Задача 8
### Рассмотрим регрессию $Y$ на $X1$, $X2$ и $X3$. Проведи тест Голдфелда–Куандта по переменной $X3$, выкидывая 40% наблюдений в середине. В ответ надо ввести расчётное значение статистики.

Рассматривается модель:

$$
Y_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i + \varepsilon_i
$$

### Теория

Тест Голдфелда–Куандта используется для проверки гипотезы о гомоскедастичности:

$$
H_0: Var(\varepsilon_i) = \sigma^2
$$

$$
H_1: Var(\varepsilon_i) \text{ зависит от } X3
$$

Алгоритм теста:

1. Отсортировать наблюдения по переменной $X3$  
2. Удалить центральную часть выборки (в данной задаче — 40%)  
3. Разделить оставшиеся наблюдения на две группы (с низкими и высокими значениями $X3$)  
4. Оценить регрессии отдельно для двух групп  
5. Посчитать статистику:

$$
F = \frac{RSS_2 / df_2}{RSS_1 / df_1}
$$

где:
- $RSS_1$, $RSS_2$ — суммы квадратов остатков в двух подвыборках  

Если $F$ сильно отличается от 1, это свидетельствует о гетероскедастичности.

In [23]:
df_sorted = df.sort_values("X3").reset_index(drop=True)

n = len(df_sorted)
drop_n = int(round(n * 0.40))
remain_n = n - drop_n
group_n = remain_n // 2

low = df_sorted.iloc[:group_n].copy()
high = df_sorted.iloc[n - group_n:].copy()

gq_low = smf.ols("Y ~ X1 + X2 + X3", data=low).fit()
gq_high = smf.ols("Y ~ X1 + X2 + X3", data=high).fit()

rss_low = np.sum(np.asarray(gq_low.resid) ** 2)
rss_high = np.sum(np.asarray(gq_high.resid) ** 2)

df_low = int(gq_low.df_resid)
df_high = int(gq_high.df_resid)

round((rss_high / df_high) / (rss_low / df_low), 3)

np.float64(1.305)

### Интерпретация

Полученная статистика теста Голдфелда–Куандта равна

$$
F = 1.305
$$

Значение близко к 1, что означает, что дисперсии остатков в группах с низкими и высокими значениями $X3$ практически не отличаются.

Следовательно, **нет признаков гетероскедастичности по переменной $X3$**, и гипотеза о гомоскедастичности не отвергается.

---
# Задача 9
### Оцени регрессию $Y$ на $X1$ и $X2$. Проведи тест Глейзера на гетероскедастичность, предполагая, что модуль остатков зависит только от квадрата $X2$. В ответ введи p-value.

Рассматривается модель:

$$
Y_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \varepsilon_i
$$

### Теория

Тест Глейзера используется для проверки гетероскедастичности, предполагая, что дисперсия ошибок зависит от некоторой функции регрессоров.

Алгоритм теста:

1. Оценить исходную регрессию и получить остатки $\hat{\varepsilon}_i$  
2. Взять их модуль: $|\hat{\varepsilon}_i|$  
3. Оценить вспомогательную регрессию:

$$
|\hat{\varepsilon}_i| = \gamma_0 + \gamma_1 X2_i^2 + u_i
$$

4. Проверить гипотезу:

$$
H_0: \gamma_1 = 0 \quad (\text{гомоскедастичность})
$$

$$
H_1: \gamma_1 \ne 0 \quad (\text{гетероскедастичность})
$$

Если коэффициент при $X2^2$ значим (малое p-value), это свидетельствует о гетероскедастичности.

In [13]:
X = sm.add_constant(df[["X1", "X2"]])
y = df["Y"]

model = sm.OLS(y, X).fit()
residuals = model.resid.abs()

df["X2_sq"] = df["X2"] ** 2

X_aux = sm.add_constant(df[["X2_sq"]])
model_aux = sm.OLS(residuals, X_aux).fit()

model_aux.pvalues["X2_sq"]

np.float64(0.09247064042815228)

### Интерпретация

Полученное значение

$$
p\text{-value} = 0.0925
$$

означает, что на стандартном уровне значимости $5\%$ нулевая гипотеза о гомоскедастичности **не отвергается**, поскольку

$$
0.0925 > 0.05
$$

Но на уровне значимости $10\%$ гипотеза уже отвергалась бы, так как

$$
0.0925 < 0.10
$$

Следовательно, есть только **слабые признаки гетероскедастичности**: зависимость модуля остатков от $X2^2$ не выглядит убедительной на уровне $5\%$, но может считаться пограничной на уровне $10\%$.

---
# Задача 10
### Рассмотрим регрессию $Y$ на $X1$, $X2$ и $X3$. Рассчитай $t$-статистику для гипотезы о незначимости коэффициента при $X2$, используя робастные ошибки в форме Cribari-Neto (2004).

Рассматривается модель:

$$
Y_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i + \varepsilon_i
$$

Нужно проверить гипотезу

$$
H_0: \beta_2 = 0
$$

против альтернативы

$$
H_1: \beta_2 \ne 0
$$

### Теория

В лекциях для случая гетероскедастичности рассматриваются робастные стандартные ошибки Уайта и их варианты HC0, HC1, HC2, HC3. :contentReference[oaicite:0]{index=0}

В задаче просят использовать робастные ошибки **Cribari-Neto (2004)**. Это соответствует оценке типа **HC4**, которая специально учитывает возможное влияние наблюдений с большим leverage. :contentReference[oaicite:1]{index=1}

Общая форма робастной ковариационной матрицы:

$$
\widehat{Var}(\hat\beta) = (X^\top X)^{-1} X^\top \Omega X (X^\top X)^{-1}
$$

Для HC4 диагональные элементы матрицы $\Omega$ задаются как

$$
\omega_i = \frac{\hat\varepsilon_i^2}{(1-h_{ii})^{\delta_i}},
\qquad
\delta_i = \min\left(4, \frac{n h_{ii}}{p}\right),
$$

где:
- $\hat\varepsilon_i$ — остаток,
- $h_{ii}$ — диагональный элемент hat matrix,
- $n$ — число наблюдений,
- $p$ — число параметров модели, включая константу. :contentReference[oaicite:2]{index=2}

После этого стандартная ошибка коэффициента при $X2$ равна

$$
se(\hat\beta_2) = \sqrt{\widehat{Var}(\hat\beta)_{22}}
$$

и искомая статистика вычисляется по формуле

$$
t = \frac{\hat\beta_2}{se(\hat\beta_2)}
$$

Поскольку в `statsmodels` встроены только HC0–HC3, а HC4 в документации не заявлен, здесь удобнее посчитать её вручную. :contentReference[oaicite:3]{index=3}

In [14]:
X = sm.add_constant(df[["X1", "X2", "X3"]])
y = df["Y"]

model = sm.OLS(y, X).fit()

X_mat = X.to_numpy()
resid = model.resid.to_numpy()
hat = model.get_influence().hat_matrix_diag

n = X_mat.shape[0]
p = X_mat.shape[1]
delta = np.minimum(4, n * hat / p)
omega = resid**2 / (1 - hat) ** delta

XtX_inv = np.linalg.inv(X_mat.T @ X_mat)
cov_hc4 = XtX_inv @ (X_mat.T @ np.diag(omega) @ X_mat) @ XtX_inv

se_hc4 = np.sqrt(np.diag(cov_hc4))
t_x2 = model.params["X2"] / se_hc4[list(X.columns).index("X2")]

t_x2

np.float64(2.702993585040607)

### Интерпретация

Полученная статистика равна

$$
t = 2.703
$$

Она используется для проверки гипотезы

$$
H_0: \beta_{X2} = 0
$$

против альтернативы

$$
H_1: \beta_{X2} \ne 0
$$

По модулю это значение заметно больше 2, поэтому коэффициент при $X2$ остается **статистически значимым** даже при использовании робастных стандартных ошибок HC4 (Cribari-Neto, 2004).

Это означает, что вывод о значимости переменной $X2$ **устойчив к возможной гетероскедастичности**.

---
# Задача 11
### Рассмотрим пробит-регрессию $Y\_bin$ на $X1$, $X2$ и $X3$. В ответ введи коэффициент при $X2$ в этой регрессии

Рассматривается модель бинарного выбора:

$$
P(Y_i = 1 \mid X_i) = F(\beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i)
$$

где в пробит-модели $F(\cdot)$ — функция распределения стандартного нормального закона. Это прямо разбирается в лекции по моделям бинарного выбора. :contentReference[oaicite:0]{index=0}

### Теория

Для бинарной зависимой переменной линейная модель вероятности имеет недостатки: предсказанные значения могут выходить за интервал от 0 до 1, а дисперсия ошибки зависит от $X$. Поэтому используют нелинейные модели бинарного выбора, в частности логит и пробит. :contentReference[oaicite:1]{index=1}

В пробит-модели предполагается существование латентной переменной:

$$
Y_i^* = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i + \varepsilon_i
$$

и наблюдаемая бинарная переменная задаётся так:

$$
Y_i =
\begin{cases}
1, & \text{если } Y_i^* \ge 0 \\
0, & \text{если } Y_i^* < 0
\end{cases}
$$

Параметры пробит-модели оцениваются методом максимального правдоподобия. :contentReference[oaicite:2]{index=2}

В задаче нужно оценить пробит-регрессию $Y\_bin$ на $X1$, $X2$, $X3$ и взять коэффициент при $X2$.

In [15]:
X = sm.add_constant(df[["X1", "X2", "X3"]])
y = df["Y_bin"]

model = sm.Probit(y, X).fit(disp=False)
model.params["X2"]

np.float64(-0.32932463801387885)

### Интерпретация

Оценённый коэффициент равен

$$
\hat{\beta}_{X2} = -0.329
$$

Это означает, что увеличение переменной $X2$ приводит к **снижению вероятности того, что $Y\_{bin} = 1$**.

Важно, что в пробит-модели коэффициенты напрямую не интерпретируются в вероятностях. Знак коэффициента показывает направление влияния:

- отрицательный знак ⇒ вероятность уменьшается  
- положительный знак ⇒ вероятность увеличивается  

Таким образом, $X2$ оказывает **отрицательное влияние** на вероятность наступления события $Y=1$.

---
# Задача 12
### Чему равно отношение шансов (Odds Ratio) для переменной $X2$ в логистической регрессии $Y\_{bin}$ на $X1$, $X2$ и $X3$?

Рассматривается логит-модель:

$$
P(Y_i = 1) = \frac{1}{1 + e^{-Z_i}}, \quad Z_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i
$$

### Теория

В логит-модели коэффициенты интерпретируются через **отношение шансов (odds ratio)**.

Шансы определяются как:

$$
Odds = \frac{P(Y=1)}{P(Y=0)}
$$

Логарифм шансов:

$$
\ln(Odds) = \beta_0 + \beta_1 X1 + \beta_2 X2 + \beta_3 X3
$$

При увеличении $X2$ на 1 единицу:

$$
Odds \text{ изменяется в } e^{\beta_2} \text{ раз}
$$

Таким образом, отношение шансов для $X2$ равно:

$$
OR = e^{\beta_2}
$$

и показывает, во сколько раз изменяются шансы при увеличении $X2$ на единицу. :contentReference[oaicite:0]{index=0}

In [18]:
X = sm.add_constant(df[["X1", "X2", "X3"]])
y = df["Y_bin"]

model = sm.Logit(y, X).fit(disp=False)
np.exp(model.params["X2"])

np.float64(0.5736131146567719)

### Интерпретация

Полученное значение

$$
OR = 0.574
$$

Это означает, что при увеличении $X2$ на 1 единицу **шансы наступления события $Y=1$ уменьшаются примерно в 0.574 раза**.

Так как $OR < 1$, влияние переменной $X2$ является отрицательным: рост $X2$ снижает вероятность наступления события.

---
# Задача 13
### В пробит-регрессии $Y\_{bin}$ на $X1$, $X2$ и $X3$ проверьте совместную значимость $X1$ и $X3$ при помощи теста Вальда. В ответ введите p-value

Рассматривается модель:

$$
P(Y_i = 1) = F(\beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i)
$$

### Теория

Тест Вальда используется для проверки гипотез о **нескольких линейных ограничениях** на параметры модели.

В лекции отмечается, что тест Вальда применяется для проверки гипотез вида: :contentReference[oaicite:0]{index=0}

$$
H_0: R\beta = r
$$

В данной задаче:

$$
H_0: \beta_1 = \beta_3 = 0
$$

$$
H_1: \text{хотя бы один из коэффициентов } \ne 0
$$

Статистика Вальда имеет асимптотическое распределение $\chi^2$.

Если p-value маленькое, гипотеза $H_0$ отвергается ⇒ переменные совместно значимы.

In [19]:
X = sm.add_constant(df[["X1", "X2", "X3"]])
y = df["Y_bin"]

model = sm.Probit(y, X).fit(disp=False)

R = np.array([
    [0, 1, 0, 0],
    [0, 0, 0, 1]
])

wald_test = model.wald_test(R)
round(float(wald_test.pvalue), 3)

/Users/maria/Desktop/Code/.venv/lib/python3.12/site-packages/statsmodels/base/model.py:1912: FutureWarning: The behavior of wald_test will change after 0.14 to returning scalar test statistic values. To get the future behavior now, set scalar to True. To silence this message while retaining the legacy behavior, set scalar to False.
  warnings.warn(


0.0

### Интерпретация

Полученное значение

$$
p\text{-value} = 0.000
$$

(с точностью до округления)

означает, что нулевая гипотеза

$$
H_0: \beta_{X1} = \beta_{X3} = 0
$$

уверенно отвергается на любых стандартных уровнях значимости.

Следовательно, переменные $X1$ и $X3$ **совместно статистически значимы** в пробит-модели и вносят вклад в объяснение вероятности $Y=1$.

---
# Задача 14
### В логистической регрессии $Y\_{bin}$ на $X1$, $X2$ и $X3$ рассчитайте предельные эффекты в точке $X1 = 10$, $X2 = 2$, $X3 = 2$. В ответ введите предельный эффект по переменной $X1$

Рассматривается логит-модель:

$$
P(Y_i = 1) = \frac{1}{1 + e^{-Z_i}}, \quad Z_i = \beta_0 + \beta_1 X1_i + \beta_2 X2_i + \beta_3 X3_i
$$

### Теория

Предельный эффект показывает, как изменяется вероятность при изменении объясняющей переменной:

$$
\frac{\partial P}{\partial X_j} = f(Z) \cdot \beta_j
$$

где:

$$
f(Z) = \frac{e^{-Z}}{(1 + e^{-Z})^2}
$$

— производная логистической функции.

В отличие от линейной модели, предельный эффект **зависит от точки**, в которой он вычисляется.

В задаче нужно:
1. Посчитать $Z$ в точке $(X1, X2, X3) = (10, 2, 2)$  
2. Найти $f(Z)$  
3. Умножить на коэффициент при $X1$  

$$
ME_{X1} = f(Z) \cdot \beta_1
$$

In [20]:
X = sm.add_constant(df[["X1", "X2", "X3"]])
y = df["Y_bin"]

model = sm.Logit(y, X).fit(disp=False)

b = model.params

Z = b["const"] + b["X1"]*10 + b["X2"]*2 + b["X3"]*2

f = np.exp(-Z) / (1 + np.exp(-Z))**2

round(float(f * b["X1"]), 3)

0.0

### Интерпретация

Полученное значение предельного эффекта равно

$$
ME_{X1} = 0.000
$$

(с точностью до округления)

Это означает, что в заданной точке $(X1, X2, X3) = (10, 2, 2)$ вероятность $Y=1$ **практически не изменяется при изменении $X1$**.

Такое возможно, когда значение $Z$ находится в области, где логистическая функция «насыщается» (близко к 0 или 1), и её производная $f(Z)$ становится очень маленькой.

---
# Задача 15
### Будем считать, что пробит-регрессия $Y\_{bin}$ на $X1$, $X2$ и $X3$ предсказывает 1, когда оценка вероятности по модели больше либо равна 0.7 (и 0, когда меньше 0.7). Какую долю исходов правильно предсказала пробит-регрессия?

### Теория

После оценки пробит-модели получаем вероятности:

$$
\hat{P}_i = P(Y_i = 1 \mid X_i)
$$

Далее строится правило классификации:

$$
\hat{Y}_i =
\begin{cases}
1, & \hat{P}_i \ge 0.7 \\
0, & \hat{P}_i < 0.7
\end{cases}
$$

Качество прогноза оценивается через **долю правильно классифицированных наблюдений**:

$$
Accuracy = \frac{\text{число верных предсказаний}}{n}
$$

То есть доля случаев, когда $\hat{Y}_i = Y_i$.

In [22]:
X = sm.add_constant(df[["X1", "X2", "X3"]])
y = df["Y_bin"]

model = sm.Probit(y, X).fit(disp=False)

probs = model.predict(X)
y_pred = (probs >= 0.7).astype(int)

round(float((y_pred == y).mean()), 3)

0.74

### Интерпретация

Доля правильно предсказанных исходов равна

$$
0.740
$$

Это означает, что пробит-модель правильно классифицировала **74.0\% наблюдений** при пороге классификации $0.7$.

Иными словами, в 74 случаях из 100 модель верно определила значение бинарной переменной $Y_{bin}$.

Это можно считать **достаточно неплохим качеством классификации**, хотя итоговая оценка качества зависит и от баланса классов, и от выбранного порога $0.7$.

еще одно решение для сверки

In [17]:
import os
import numpy as np
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
from scipy import stats
from statsmodels.stats.outliers_influence import variance_inflation_factor

# ============================================================
# ЗАГРУЗКА ДАННЫХ
# ============================================================


unnamed_cols = [c for c in df.columns if str(c).startswith("Unnamed:")]
if unnamed_cols:
    df = df.drop(columns=unnamed_cols)

expected_cols = {"Y", "Y_bin", "X1", "X2", "X3"}
if not expected_cols.issubset(set(df.columns)):
    raise ValueError(
        f"В данных должны быть колонки {expected_cols}, а сейчас: {list(df.columns)}"
    )

# ============================================================
# ВСПОМОГАТЕЛЬНЫЕ ФУНКЦИИ
# ============================================================
def add_const(x: pd.DataFrame) -> pd.DataFrame:
    return sm.add_constant(x, has_constant="add")


def hc4_se(ols_result):
    """
    Ручной расчет HC4 ковариационной матрицы.
    HC4 = Cribari-Neto (2004).
    """
    X = np.asarray(ols_result.model.exog, dtype=float)
    resid = np.asarray(ols_result.resid, dtype=float)

    n, k = X.shape
    XtX_inv = np.linalg.inv(X.T @ X)
    hat_diag = np.sum(X * (X @ XtX_inv), axis=1)

    delta = np.minimum(4.0, n * hat_diag / k)
    omega = (resid ** 2) / ((1.0 - hat_diag) ** delta)

    S = np.zeros((k, k), dtype=float)
    for i in range(n):
        xi = X[i : i + 1].T
        S += omega[i] * (xi @ xi.T)

    cov_hc4 = XtX_inv @ S @ XtX_inv
    se_hc4 = np.sqrt(np.diag(cov_hc4))
    return cov_hc4, se_hc4


# ============================================================
# ЗАДАЧА 1
# ============================================================
df["X2_cube"] = df["X2"] ** 3
model_1 = smf.ols("Y ~ X1 + X2 + X3 + X2_cube", data=df).fit()
ans_1 = model_1.rsquared
print(f"Ответ к заданию 1: {ans_1:.3f}")

# ============================================================
# ЗАДАЧА 2
# ============================================================
df["X1_cube"] = df["X1"] ** 3
df["X3_cube"] = df["X3"] ** 3

m1 = smf.ols("Y ~ X1 + X2 + X3 + X1_cube", data=df).fit()
m2 = smf.ols("Y ~ X1 + X2 + X3 + X2_cube", data=df).fit()
m3 = smf.ols("Y ~ X1 + X2 + X3 + X3_cube", data=df).fit()

adj_r2_list = [m1.rsquared_adj, m2.rsquared_adj, m3.rsquared_adj]
ans_2 = int(np.argmax(adj_r2_list) + 1)
print(f"Ответ к заданию 2: {ans_2}")

# ============================================================
# ЗАДАЧА 3
# ============================================================
model_3 = smf.ols("Y ~ X1 + X2 + X3", data=df).fit()
f_test_3 = model_3.f_test("X1 = 0, X2 = 0")
ans_3 = float(np.asarray(f_test_3.fvalue).item())
print(f"Ответ к заданию 3: {ans_3:.3f}")

# ============================================================
# ЗАДАЧА 4
# ============================================================
df["D"] = (df["X3"] > df["X3"].mean()).astype(int)
model_4 = smf.ols("Y ~ D", data=df).fit()
ans_4 = model_4.params["D"]
print(f"Ответ к заданию 4: {ans_4:.3f}")

# ============================================================
# ЗАДАЧА 5
# ============================================================
corr_matrix = df[["X1", "X2", "X3"]].corr().abs()
np.fill_diagonal(corr_matrix.values, np.nan)
ans_5 = np.nanmax(corr_matrix.values)
print(f"Ответ к заданию 5: {ans_5:.3f}")

# ============================================================
# ЗАДАЧА 6
# ============================================================
X_vif = add_const(df[["X1", "X2", "X3"]])
vif_values = {
    X_vif.columns[i]: variance_inflation_factor(X_vif.values, i)
    for i in range(X_vif.shape[1])
}
ans_6 = vif_values["X2"]
print(f"Ответ к заданию 6: {ans_6:.3f}")

# ============================================================
# ЗАДАЧА 7
# ============================================================
if (df["Y"] <= 0).any():
    raise ValueError("В данных есть Y <= 0, логарифм Y не определен.")
if (df["X3"] <= 0).any():
    raise ValueError("В данных есть X3 <= 0, логарифм X3 не определен.")

df["logY"] = np.log(df["Y"])
df["logX3"] = np.log(df["X3"])
model_7 = smf.ols("logY ~ logX3", data=df).fit()
ans_7 = model_7.params["logX3"]
print(f"Ответ к заданию 7: {ans_7:.3f}")

# ============================================================
# ЗАДАЧА 8
# ============================================================
df_sorted = df.sort_values("X3").reset_index(drop=True)

n = len(df_sorted)
drop_n = int(round(n * 0.40))
remain_n = n - drop_n
group_n = remain_n // 2

low = df_sorted.iloc[:group_n].copy()
high = df_sorted.iloc[n - group_n :].copy()

gq_low = smf.ols("Y ~ X1 + X2 + X3", data=low).fit()
gq_high = smf.ols("Y ~ X1 + X2 + X3", data=high).fit()

rss_low = np.sum(np.asarray(gq_low.resid) ** 2)
rss_high = np.sum(np.asarray(gq_high.resid) ** 2)

df_low = int(gq_low.df_resid)
df_high = int(gq_high.df_resid)

ans_8 = (rss_high / df_high) / (rss_low / df_low)
print(f"Ответ к заданию 8: {ans_8:.3f}")

# ============================================================
# ЗАДАЧА 9
# ============================================================
model_9_main = smf.ols("Y ~ X1 + X2", data=df).fit()
df["abs_resid_9"] = np.abs(np.asarray(model_9_main.resid))
df["X2_sq"] = df["X2"] ** 2

model_9_aux = smf.ols("abs_resid_9 ~ X2_sq", data=df).fit()
ans_9 = model_9_aux.pvalues["X2_sq"]
print(f"Ответ к заданию 9: {ans_9:.3f}")

# ============================================================
# ЗАДАЧА 10
# ============================================================
model_10 = smf.ols("Y ~ X1 + X2 + X3", data=df).fit()
cov_hc4, se_hc4 = hc4_se(model_10)

param_names = model_10.model.exog_names
idx_x2 = param_names.index("X2")

beta_x2 = model_10.params["X2"]
ans_10 = beta_x2 / se_hc4[idx_x2]
print(f"Ответ к заданию 10: {ans_10:.3f}")

# ============================================================
# ЗАДАЧА 11
# ============================================================
probit_model = smf.probit("Y_bin ~ X1 + X2 + X3", data=df).fit(disp=False)
ans_11 = probit_model.params["X2"]
print(f"Ответ к заданию 11: {ans_11:.3f}")

# ============================================================
# ЗАДАЧА 12
# ============================================================
logit_model = smf.logit("Y_bin ~ X1 + X2 + X3", data=df).fit(disp=False)
ans_12 = np.exp(logit_model.params["X2"])
print(f"Ответ к заданию 12: {ans_12:.3f}")

# ============================================================
# ЗАДАЧА 13
# ============================================================
wald_13 = probit_model.wald_test("X1 = 0, X3 = 0")
ans_13 = float(np.asarray(wald_13.pvalue).item())
print(f"Ответ к заданию 13: {ans_13:.3f}")

# ============================================================
# ЗАДАЧА 14
# ============================================================
b = logit_model.params
z = b["Intercept"] + b["X1"] * 10 + b["X2"] * 2 + b["X3"] * 2
p = 1.0 / (1.0 + np.exp(-z))
ans_14 = p * (1.0 - p) * b["X1"]
print(f"Ответ к заданию 14: {ans_14:.3f}")

# ============================================================
# ЗАДАЧА 15
# ============================================================
pred_prob_probit = np.asarray(probit_model.predict(df))
pred_class = (pred_prob_probit >= 0.7).astype(int)
ans_15 = np.mean(pred_class == np.asarray(df["Y_bin"]))
print(f"Ответ к заданию 15: {ans_15:.3f}")

Ответ к заданию 1: 0.642
Ответ к заданию 2: 1
Ответ к заданию 3: 10.980
Ответ к заданию 4: 207.980
Ответ к заданию 5: 0.183
Ответ к заданию 6: 1.042
Ответ к заданию 7: 0.733
Ответ к заданию 8: 1.305
Ответ к заданию 9: 0.092
Ответ к заданию 10: 2.703
Ответ к заданию 11: -0.329
Ответ к заданию 12: 0.574
Ответ к заданию 13: 0.000
Ответ к заданию 14: 0.000
Ответ к заданию 15: 0.740


/Users/maria/Desktop/Code/.venv/lib/python3.12/site-packages/statsmodels/base/model.py:1912: FutureWarning: The behavior of wald_test will change after 0.14 to returning scalar test statistic values. To get the future behavior now, set scalar to True. To silence this message while retaining the legacy behavior, set scalar to False.
  warnings.warn(


еще одно решение для сверки

In [24]:
from scipy.special import expit
from statsmodels.discrete.discrete_model import Logit, Probit
from statsmodels.stats.diagnostic import het_goldfeldquandt
from statsmodels.stats.outliers_influence import OLSInfluence, variance_inflation_factor


def fit_ols(y, x_cols):
    return sm.OLS(y, sm.add_constant(df[x_cols])).fit()


model_xyz = fit_ols(df['Y'], ['X1', 'X2', 'X3'])

model_x1_cube = sm.OLS(
    df['Y'],
    sm.add_constant(df[['X1', 'X2', 'X3']].assign(X1_cubed=df['X1'] ** 3)),
).fit()
model_x2_cube = sm.OLS(
    df['Y'],
    sm.add_constant(df[['X1', 'X2', 'X3']].assign(X2_cubed=df['X2'] ** 3)),
).fit()
model_x3_cube = sm.OLS(
    df['Y'],
    sm.add_constant(df[['X1', 'X2', 'X3']].assign(X3_cubed=df['X3'] ** 3)),
).fit()

D = (df['X3'] > df['X3'].mean()).astype(int).rename('D')
X_with_const = sm.add_constant(df[['X1', 'X2', 'X3']])
probit_model = Probit(df['Y_bin'], X_with_const).fit(disp=False)
logit_model = Logit(df['Y_bin'], X_with_const).fit(disp=False)

f_test_x1_x2 = model_xyz.f_test('X1 = 0, X2 = 0')
gq_stat = het_goldfeldquandt(
    df['Y'],
    X_with_const,
    idx=list(X_with_const.columns).index('X3'),
    split=0.3,
    drop=0.4,
)[0]

glejser_aux = sm.OLS(
    np.abs(fit_ols(df['Y'], ['X1', 'X2']).resid),
    sm.add_constant((df['X2'] ** 2).rename('X2_sq')),
).fit()

# Cribari-Neto (2004) HC4 covariance matrix.
hat_diag = OLSInfluence(model_xyz).hat_matrix_diag
x_np = X_with_const.to_numpy()
xtx_inv = np.linalg.inv(x_np.T @ x_np)
delta = np.minimum(4, hat_diag / hat_diag.mean())
omega = model_xyz.resid.to_numpy() ** 2 / (1 - hat_diag) ** delta
cov_hc4 = xtx_inv @ (x_np.T @ np.diag(omega) @ x_np) @ xtx_inv
se_hc4 = pd.Series(np.sqrt(np.diag(cov_hc4)), index=X_with_const.columns)

point = pd.Series({'const': 1.0, 'X1': 10.0, 'X2': 2.0, 'X3': 2.0})
logit_prob_at_point = expit(float(point @ logit_model.params))
probit_pred = (probit_model.predict(X_with_const) >= 0.7).astype(int)

answers_raw = {
    1: model_x2_cube.rsquared,
    2: max(
        {1: model_x1_cube.rsquared_adj, 2: model_x2_cube.rsquared_adj, 3: model_x3_cube.rsquared_adj},
        key=lambda key: {1: model_x1_cube.rsquared_adj, 2: model_x2_cube.rsquared_adj, 3: model_x3_cube.rsquared_adj}[key],
    ),
    3: float(f_test_x1_x2.fvalue),
    4: sm.OLS(df['Y'], sm.add_constant(D)).fit().params['D'],
    5: df[['X1', 'X2', 'X3']].corr().abs().where(~np.eye(3, dtype=bool)).max().max(),
    6: variance_inflation_factor(x_np, list(X_with_const.columns).index('X2')),
    7: sm.OLS(np.log(df['Y']), sm.add_constant(np.log(df['X3']).rename('log_X3'))).fit().params['log_X3'],
    8: gq_stat,
    9: glejser_aux.pvalues['X2_sq'],
    10: model_xyz.params['X2'] / se_hc4['X2'],
    11: probit_model.params['X2'],
    12: np.exp(logit_model.params['X2']),
    13: float(probit_model.wald_test('X1 = 0, X3 = 0', scalar=True).pvalue),
    14: logit_model.params['X1'] * logit_prob_at_point * (1 - logit_prob_at_point),
    15: (probit_pred == df['Y_bin']).mean(),
}

answers = pd.Series(
    {
        task: str(value) if task == 2 else f'{float(value):.3f}'
        for task, value in answers_raw.items()
    },
    name='answer',
)
answers

1       0.642
2           1
3      10.980
4     207.980
5       0.183
6       1.042
7       0.733
8       1.305
9       0.092
10      2.703
11     -0.329
12      0.574
13      0.000
14      0.000
15      0.740
Name: answer, dtype: object

In [25]:
# ============================================================
# ЗАДАЧА 13
# В probit-регрессии Y_bin на X1, X2, X3
# проверить совместную значимость X1 и X3 при помощи теста Вальда
# H0: beta_X1 = 0, beta_X3 = 0
# ============================================================

probit_model = smf.probit("Y_bin ~ X1 + X2 + X3", data=df).fit(disp=False)

wald_13 = probit_model.wald_test("X1 = 0, X3 = 0", use_f=False)
ans_13 = float(np.asarray(wald_13.pvalue).item())

print(f"Ответ к заданию 13: {ans_13:.3f}")

Ответ к заданию 13: 0.000


/Users/maria/Desktop/Code/.venv/lib/python3.12/site-packages/statsmodels/base/model.py:1912: FutureWarning: The behavior of wald_test will change after 0.14 to returning scalar test statistic values. To get the future behavior now, set scalar to True. To silence this message while retaining the legacy behavior, set scalar to False.
  warnings.warn(
